Sources/NIOPosix/DynamicResolver.swift

```diff
@@ -0,0 +1,64 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2023 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import NIOCore
+import Atomics
+
+/// Wraps a use-exactly-once resolver and manages its lifecycle, allowing it to be virtually used never or more than once.
+/// Each time a complete resolution happen, meaning that both the A and a AAAA queries are initiated, the current resolver is replaced by a new one.
+///
+/// This can be useful in situations where it's not possible to assure that the resolver's single-use nature is honored.
+class DynamicResolver<T: Resolver>: Resolver {
+    private let resolverFactory: () -> T
+    private var resolver: T
+    
+    // The status of the resolver is tracked by this atomic integer and can assume the following values:
+    // * 0: Resolver has been initialized but has not been used.
+    // * 1: Resolver has initiated one query.
+    // * 2: Resolver has initiated the second and last query, time to refresh it.
+    private let resolverStatus: ManagedAtomic<Int>
+    
+    init(resolverFactory: @escaping () -> T) {
+        self.resolverFactory = resolverFactory
+        resolver = resolverFactory()
+        resolverStatus = ManagedAtomic(0)
+    }
+    
+    func initiateAQuery(host: String, port: Int) -> NIOCore.EventLoopFuture<[NIOCore.SocketAddress]> {
```

Comment:
I don't think these implementations are correct in the face of multiple concurrent invocations. We probably want to defend against that. We may actually need a new protocol to pull that off.